### PR TITLE
Synthesize Decoupled from IFluidObject and Deprecations Removed

### LINKED ISSUE #8631: Remove Synthesize Deprecated
Once we remove deprecated on IFluidDependecySynthsize, and DependencyContainer we should do the following:

yeah. there is another set of changes coming. once i no longer expose providers directly i can just create the `PropertyDescriptions` upfront on register, and then all this code will be way simpler. I do those change with what you suggest in the next release

_Originally posted by @anthony-murphy in https://github.com/microsoft/FluidFramework/pull/8593#r776830966_

### PR DESCRIPTION
**Pull request by anthony-murphy**
DependencyContainer now takes a generic argument, as it is no longer directly couple to IFluidObject. The ideal pattern here would be directly pass the provider or FluidObject interfaces you will register. As a short term solution you could also pass IFluidObject, but IFluidObject is deprecated, so will need to be removed if used here. 
Examples:
``` typescript
    // the old way
    const dc = new DependencyContainer();
    dc.register(IFluidHTMLView, MockLoadable());
    // FluidObject option
    const dc = new DependencyContainer<FluidObject<IFluidHTMLView>>();
    dc.register(IFluidHTMLView, MockLoadable());
    // Provider option
    const dc = new DependencyContainer<IProvideFluidHTMLView>();
    dc.register(IFluidHTMLView, MockLoadable());
    // Short term IFluidObject option
    const dc = new DependencyContainer<IFluidObject>();
    dc.register(IFluidHTMLView, MockLoadable());
```

The following members have been removed from IFluidDependencySynthesizer:
 - registeredTypes - unused and no longer supported. `has` can replace most possible usages
 - register - create new DependencyContainer and add existing as parent
 - unregister - create new DependencyContainer and add existing as parent
 - getProvider - use `has` and `synthesize` to check or get provider respectively 

 The following types have been removed or changed. These changes should only affect direct usages which should be rare. Existing synthesizer api usage is backwards compatible:
 - FluidObjectKey - removed as IFluidObject is deprecated
 - NonNullableFluidObject - removed as IFluidObject is deprecated. use typescripts NonNullable instead
 - AsyncRequiredFluidObjectProvider - Takes FluidObject types rather than keys
 - AsyncOptionalFluidObjectProvider - Takes FluidObject types rather than keys
 - AsyncFluidObjectProvider - Takes FluidObject types rather than keys
 - FluidObjectProvider - Takes FluidObject types rather than keys
 - ProviderEntry - no longer used
 - DependencyContainerRegistry - no longer used

fixes #8631 
related #8074 #8852

### CODE REVIEW

File: BREAKING.md [hunk 1]
```diff
@@ -24,6 +25,48 @@ The `sharingLinkToRedeem` property is removed from the `IOdspResolvedUrl` interf
...
+
+```
+
+The following members have been removed from IFluidDependencySynthesizer:
+ - registeredTypes - unused and no longer supported. `has` can replace most possible usages
+ - register - create new DependencyContainer and add existing as parent
+ - unregister - create new DependencyContainer and add existing as parent
+ - getProvider - use `has` and `synthesize` to check or get provider respectively 
+
+ The following types have been removed or changed. These changes should only affect direct usages which should be rare. Exiting synthesizer api usage is backwards compatible:
```
Comment (anthony-murphy) [inline BREAKING.md:56]:
```suggestion
 The following types have been removed or changed. These changes should only affect direct usages which should be rare. Existing synthesizer api usage is backwards compatible:
```

File: packages/framework/synthesize/src/IFluidDependencySynthesizer.ts [hunk 2]
```diff
@@ -61,22 +30,15 @@ export interface IFluidDependencySynthesizer extends IProvideFluidDependencySynt
      * @param requiredTypes - required types that need to be in the Scope object
      */
     synthesize<
-        O extends IFluidObject,
-        R extends IFluidObject>(
+        O,
+        R = undefined  | Record<string, never>>(
```
Comment (ChumpChief) [inline packages/framework/synthesize/src/IFluidDependencySynthesizer.ts:34]: I'm not sure it's desirable to permit `undefined` here, vs. just the never record?  I'd probably prefer the stricter option to start at least.

**Review by ChumpChief — APPROVED** [review]
Mostly nits, seems good to me.

File: packages/framework/synthesize/src/IFluidDependencySynthesizer.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/framework/synthesize/src/IFluidDependencySynthesizer.ts:34]: i don't actually want the never record, but that's what we supported before. so there are existing usages that pass and emtpy object, it's better to just pass undefined, so you don't need to create and object.

**Pull request merged**